Add errorCodeToName() and system.errors

azat opened the pull request:
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- `errorCodeToName()` function - return variable name of the error (useful for analyzing query_log and similar)
- `system.errors` table - shows how many times errors has been happened (respects `system_events_show_zero_values`)

Changelog:

<details>

v2: replace hana::map/fusion::vector with external script
v3:
- use binary cmake dir for .sh
- use jump table over suboptimal std::unordered_map (with very poor hash
  function for int -- std::hash)
- cleanup errorCodeToName (drop extra templates and headers)
- rename errorCode -> errorCodeToName
- fix arcadia build (by not providing those helpers there)
- fix build on CI, by using CMAKE_CXX_COMPILER for parsing the file

</details>

HEAD

<details>

- f5ad78a284d9cf76214e8d929b644b41381d9ecc

</details>

azat committed:
Add errorCodeToName() function

This patch adds errorCode() function that will return macro name by the
exception code, useful for analyzing query_log and similar.

Later some descriptions can be added.

v2: replace hana::map/fusion::vector with external script
v3:
- use binary cmake dir for .sh
- use jump table over suboptimal std::unordered_map (with very poor hash
  function for int -- std::hash)
- cleanup errorCodeToName (drop extra templates and headers)
- rename errorCode -> errorCodeToName
- fix arcadia build (by not providing those helpers there)
- fix build on CI, by using CMAKE_CXX_COMPILER for parsing the file

alexey-milovidov commented:
errorCodeToName might be more descriptive...

alexey-milovidov commented:
> with external script

"X macro" that generates switch/case looks appropriate.
Compiler is using "jump table" for int-based switches, it's usually better than hash table.

alexey-milovidov commented:
We can also (additionally) generate `system.error_codes` table for better introspection.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+        const ColumnPtr input_column = arguments[0].column;
+        auto col_res = ColumnString::create();
+
+        if (!castTypeToEither<
```
It's heavy template instantiations, but we can use virtual call `IColumn::getUInt`.
It's suitable for the cases when performance difference does not matter.

alexey-milovidov commented on src/Functions/CMakeLists.txt:
```diff
@@ -6,6 +6,14 @@ add_headers_and_sources(clickhouse_functions .)
 list(REMOVE_ITEM clickhouse_functions_sources IFunctionImpl.cpp FunctionFactory.cpp FunctionHelpers.cpp)
 list(REMOVE_ITEM clickhouse_functions_headers IFunctionImpl.h FunctionFactory.h FunctionHelpers.h)
 
+add_custom_target(generate-error-codes
```
Arcadia build system (ya.make) theoretially supports generated source files but nobody outside of Yandex (and also no one from ClickHouse developers) can master it.

It's better to either:
- disable this function from Arcadia;
- don't use generated files;
- commit generated files to the repository.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+#    include <Common/config_version.h>
+#endif
+
+extern std::unordered_map<int, const char *> auto_error_codes;
```
Unordered map is using suboptimal hash function that will eat all the performance.

alexey-milovidov replied:
The lookup in unordered_map is also suboptimal.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+#include <unordered_map>
+
+#if !defined(ARCADIA_BUILD)
+#    include <Common/config_version.h>
```
I see no usages of this header in this file.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+        if (!isNumber(types.at(0)))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "The argument of function {} must have simple numeric type, possibly Nullable", name);
+
+        return std::make_shared<DataTypeString>();
```
Maybe consider returning LowCardinality(String)?

azat replied:
Will it be useful for the function that return single string?
Looks like extra conversion when there absolutely no point int his (if it will be added as materialized alias, then it should be wrapped with LowCardinality() - agree)

Actually I though about generating Enum for it, but this will complicate things, and I'm not sure that it worth it.

alexey-milovidov replied:
Yes. It is applicable not only for constant expressions, for example, you can process query_log with this function.
Let's do it if it's easy to do.

azat replied:
Ok, let's make it LowCardinality, but FWIW plain LowCardinality type has 2x overhead (but this is absolutely synthetic test):

- w/o low cardinality

```sql
SELECT toTypeName(errorCodeToName(0)) 
   
┌─toTypeName(errorCodeToName(0))─┐    
│ String                         │    
└────────────────────────────────┘  

SELECT errorCodeToName(number % 3000)                               
FROM system.numbers                             
LIMIT 1000000000.                                                   
FORMAT Null              

0 rows in set. Elapsed: 13.464 sec. Processed 1.00 billion rows, 8.00 GB (74.27 million rows/s., 594.20 MB/s.)    

SELECT errorCodeToName(number % 3000) 
FROM system.numbers
LIMIT 1000000000.
FORMAT Null
   
0 rows in set. Elapsed: 26.579 sec. Processed 1.00 billion rows, 8.00 GB (37.63 million rows/s., 301.01 MB/s.) 
```

- w/ LowCardinality:

```sql
SELECT toTypeName(errorCodeToName(0))

┌─toTypeName(errorCodeToName(0))─┐
│ LowCardinality(String)         │
└────────────────────────────────┘

SELECT errorCodeToName(number % 3000)
FROM system.numbers
LIMIT 1000000000.
FORMAT Null

0 rows in set. Elapsed: 26.579 sec. Processed 1.00 billion rows, 8.00 GB (37.63 million rows/s., 301.01 MB/s.) 
```

Here some profile numbers:

- w/o lowcardinality:

```
Samples: 69K of event 'cycles', 4000 Hz, Event count (approx.): 8859192325 lost: 0/0 drop: 0/0
Overhead  Shared Object        Symbol
  30.16%  clickhouse.fasttest  [.] DB::FunctionErrorCodeToName::executeImpl                                                                                                                                                                                                      ▒
  16.93%  clickhouse.fasttest  [.] DB::(anonymous namespace)::ModuloByConstantImpl<unsigned long, unsigned short>::vectorConstant                                                                                                                                                ▒
  15.82%  clickhouse.fasttest  [.] errorCodeToName                                                                                                                                                                                                                               ◆
  14.83%  clickhouse.fasttest  [.] DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::resize<>                                                                                                                                                                  ▒
   9.97%  libc-2.32.so         [.] __memmove_avx_unaligned_erms                                                                                                                                                                                                                  ▒
   3.24%  clickhouse.fasttest  [.] DB::ColumnVector<unsigned short>::getInt                                                                                                                                                                                                      ▒
   2.34%  clickhouse.fasttest  [.] DB::(anonymous namespace)::NumbersSource::generate
```

- w/o lowcardinality:

```
Samples: 97K of event 'cycles', 4000 Hz, Event count (approx.): 21324285943 lost: 0/0 drop: 0/0                                                                                                                                                                                   
Overhead  Shared Object           Symbol                                                                                                                                                                                                                                          
  19.80%  clickhouse.fasttest.lc  [.] DB::ColumnUnique<DB::ColumnString>::uniqueInsertRangeImpl<unsigned short>                                                                                                                                                                  ◆
  16.43%  clickhouse.fasttest.lc  [.] DB::ColumnString::insertData                                                                                                                                                                                                               ▒
  12.16%  clickhouse.fasttest.lc  [.] DB::ReverseIndex<unsigned long, DB::ColumnString>::insert                                                                                                                                                                                  ▒
   7.94%  clickhouse.fasttest.lc  [.] DB::FunctionErrorCodeToName::executeImpl                                                                                                                                                                                                   ▒
   7.91%  clickhouse.fasttest.lc  [.] DB::(anonymous namespace)::ModuloByConstantImpl<unsigned long, unsigned short>::vectorConstant                                                                                                                                             ▒
   7.39%  clickhouse.fasttest.lc  [.] DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 15ul, 16ul>::resize<>                                                                                                                                                               ▒
   6.62%  clickhouse.fasttest.lc  [.] DB::ErrorCodes::getName                                                                                                                                                                                                                    ▒
   5.18%  libc-2.32.so            [.] __strlen_avx2                                                                                                                                                                                                                              ▒
   4.20%  libc-2.32.so            [.] __memmove_avx_unaligned_erms                                                                                                                                                                                                               ▒
   3.61%  clickhouse.fasttest.lc  [.] memequalSSE2Wide                                                                                                                                                                                                                           ▒
   1.60%  clickhouse.fasttest.lc  [.] DB::ColumnVector<unsigned short>::getInt                                                                                                                                                                                                   ▒
   1.27%  clickhouse.fasttest.lc  [.] DB::(anonymous namespace)::NumbersSource::generate                                                                                                                                                                                         ▒

```

(Also this version has strlen, it wasn't optimized by the compiler, will address this bit)

P.S.
- benchmarks are not that accurate (use a little bit different implementations, but the profiler shows the idea) and purely synthetic
- and yes LowCardinality type can have benefits later, but still.

azat replied:
>(Also this version has strlen, it wasn't optimized by the compiler, will address this bit)

0cd79de632047a3cbecd829616abb1456ca3a96e

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
```
Unused declaration.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+/** errorCode() - returns the macro name for the error code.
```
It's not the "macro name", it's variable name.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+#    include <Common/config_version.h>
+#endif
+
+extern std::unordered_map<int, const char *> auto_error_codes;
```
Simple lookup table should be better.

azat replied:
But error codes are not sequential

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+#include <DataTypes/DataTypesNumber.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnString.h>
+#include <Core/Field.h>
```
Looks like unused header.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+                for (size_t i = 0; i < input_rows_count; ++i)
+                {
+                    const Int64 error_code = column.getInt(i);
+                    std::string macro_name;
```
Excessive allocation in a loop + extra call to strlen.
You can use std::string_view for values... but simple switch/case with string literals is Ok.

alexey-milovidov commented on src/Functions/errorCode.cpp:
```diff
@@ -0,0 +1,87 @@
+            {
+                for (size_t i = 0; i < input_rows_count; ++i)
+                {
+                    const Int64 error_code = column.getInt(i);
```
getInt will throw exception when needed, so you don't need `castTypeToEither`

alexey-milovidov commented:
PS. I totally like the idea to have this function.

azat commented:
>PS. I totally like the idea to have this function.

Yeah, this version of the patch was only to get feedback will it be helpful to someone else, so that's why it so not clean (plus I've started, a month ago, it earlier in another form, but that doesn't work out, so it contains some some bits that was used before)
Will get back with the changes, thanks for taking a look.

alexey-milovidov commented:
> We can also (additionally) generate system.error_codes table for better introspection.

We can add `system.errors` table that will collect error counts per code since server startup (similar to ProfileEvents).

azat commented:
>errorCodeToName might be more descriptive...
>Because the argument is already an "error code".

Ok (I though about this, but decided to make it less descriptive, since it returns some "mnemonic" name)

>"X macro" that generates switch/case looks appropriate.
>Compiler is using "jump table" for int-based switches, it's usually better than hash table.

Ok, (I though about simplicity for this function, rather then performance, since it should not be a hot code path)

> Also you can generate lookup table, see ProfileEvents, CurrentMetrics.

Hm, I guess it will better to do this using the same pattern, since it is already in the project (although this will make the code less readable and also there may be conflicts but this is minor)

>We can also (additionally) generate system.error_codes table for better introspection.

I though about this, but prefer to add a function instead (even though you can load system.error_codes into dictionary if you need to), since then you can add `alias exception_code_name alias errorCodeToName(exception_code)` into `query_log`  w/o extra complexity

>We can add system.errors table that will collect error counts per code since server startup (similar to ProfileEvents).

But absolutely like this idea (I though about this before), so `system.errors` (instead of `system.error_codes`) sounds great (but maybe in a separate PR, will see how it will go)

azat commented:
>Also you can generate lookup table, see ProfileEvents, CurrentMetrics.

Remember one more point against, error codes values are not sequential (548, 999), (1002, 2001) ranges are not in use, so lookup table will be suboptimal

azat committed:
Add system.errors table

Contains error codes with number of times they have been triggered.

Columns:

-   `name` ([String](../../sql-reference/data-types/string.md)) — name of the error (`errorCodeToName`).
-   `code` ([Int32](../../sql-reference/data-types/int-uint.md)) — code number of the error.
-   `value` ([UInt64](../../sql-reference/data-types/int-uint.md)) - number of times this error has been happened.

**Example**

``` sql
SELECT *
FROM system.errors
WHERE value > 0
ORDER BY code ASC
LIMIT 1

┌─name─────────────┬─code─┬─value─┐
│ CANNOT_OPEN_FILE │   76 │     1 │
└──────────────────┴──────┴───────┘

alexey-milovidov commented:
> add a function instead

Better to do both :)

> error codes values are not sequential

Does not matter, it's just about 1000 elements.

azat committed:
Make errorCodeToName() return LowCardinality(String)

azat committed:
Covert error codes list to lookup table and drop generation of them.

azat committed:
Fix readability-qualified-auto in errorCodeToName()

azat committed:
Fix readability-inconsistent-declaration-parameter-name in ErrorCodes.h

azat committed:
Make check-style happy (by not using DB::ErrorCodes:: in comments)

azat committed:
Fix ErrorCodes increment on error

azat committed:
Eliminate extra strlen() in errorCodeToName()

azat committed:
Fix readability-container-size-empty in StorageSystemErrors.cpp

azat committed:
Remove some absolute changes for errorCodeToName/system.errors

azat commented:
`Yandex synchronization check (only for Yandex employees)` fails, any clue?

azat commented:
>Yandex synchronization check (only for Yandex employees) fails, any clue?

I guess it can be a test, since errorCodeToName was disabled for the arcadia build, but the test not
But with current implementation errorCodeToName can be build for arcadia too, adjusted, will see how it goes

azat commented:
>Integration tests (release) Pending — Started

Hanged?

>Functional stateless tests (thread) — fail: 1, passed: 2384, skipped: 14

`00840_long_concurrent_select_and_drop_deadlock` does not looks related

>Integration tests (thread) — fail: 1, passed: 924, error: 0

`test_input_format_parallel_parsing_memory_tracking/test.py::test_memory_tracking_total` does not looks related

alexey-milovidov commented on src/Common/ErrorCodes.cpp:
```diff
@@ -1,526 +1,562 @@
+    M(STD_EXCEPTION, 1001) \
+    M(UNKNOWN_EXCEPTION, 1002) \
+    \
+    M(CONDITIONAL_TREE_PARENT_NOT_FOUND, 2001) \
```
BTW, 2000s codes should not exist, I will change them.

alexey-milovidov commented on src/Common/ErrorCodes.cpp:
```diff
@@ -1,526 +1,562 @@
+    M(UNKNOWN_EXCEPTION, 1002) \
+    \
+    M(CONDITIONAL_TREE_PARENT_NOT_FOUND, 2001) \
+    M(ILLEGAL_PROJECTION_MANIPULATOR, 2002) \
```
BTW, if we change from 
UNKNOWN_VOLUME, 548
to
548, UNKNOWN_VOLUME
the code will be slightly more readable.

alexey-milovidov commented on src/Common/ErrorCodes.cpp:
```diff
@@ -1,526 +1,562 @@
-    extern const int UNKNOWN_VOLUME = 548;
+    struct ErrorCodesNames
+    {
+        std::string_view names[END+1];
```
Array initialization will be totally Ok (no need for constructor).

alexey-milovidov commented on src/Common/ErrorCodes.h:
```diff
@@ -0,0 +1,41 @@
+    /// Add value for specified error_code.
+    inline void increment(ErrorCode error_code)
+    {
+        error_code = std::min<Value>(error_code, end()-1);
```
Why do we need this?

azat replied:
To avoid overflow, i.e. for `increment(-1)`

azat replied:
But `Value` is int, so this works incorrectly, fixed (and make this explicit)

alexey-milovidov commented on src/Storages/System/StorageSystemErrors.h:
```diff
@@ -0,0 +1,29 @@
+class Context;
+
+
+/** Implements the `distribution_queue` system table, which allows you to view the INSERT queues for the Distributed tables.
```
Outdated comment.

alexey-milovidov commented on src/Common/ErrorCodes.h:
```diff
@@ -0,0 +1,41 @@
+namespace ErrorCodes
+{
+    /// ErrorCode identifier (index in array).
+    using ErrorCode = size_t;
```
It looks strange.

azat replied:
This was done by analogy from the CurrentMetrics.h and similar

azat committed:
Update comment for StorageSystemErrors

azat committed:
Swap name and value in ErrorCodes definition

azat committed:
Fix overflow check in ErrorCodes

alexey-milovidov committed:
Merge branch 'master' into errorCode-v2

azat committed:
Fix 01339_client_unrecognized_option after UNRECOGNIZED_ARGUMENTS value change

alexey-milovidov commented on tests/queries/0_stateless/01544_errorCodeToName.sql:
```diff
@@ -0,0 +1,4 @@
+SELECT errorCodeToName(toUInt32(-1));
+SELECT errorCodeToName(600); /* gap in error codes */
```
A time bomb.

alexey-milovidov replied:
It's Ok.

alexey-milovidov commented:
Next steps:
- add time of the last error;
- add exception message of the last error;
- add stack trace of the last error;

filimonov commented:
Also - respect system_metrics_events_show_zero_values (or introduce new settings).

azat commented:
>Also - respect system_metrics_events_show_zero_values (or introduce new setting).

@filimonov There is no `system_metrics_events_show_zero_values`, but it does respect the `system_events_show_zero_values`

azat commented:
>Next steps:
>
>add time of the last error;
>add exception message of the last error;
>add stack trace of the last error;

#21529 (+ `remote` column, introduced in #19290)
